src/lsp/cobol_parser/expect/gen_expect.ml

```
@@ -0,0 +1,421 @@
+      | s when CompEntrySet.is_empty s -> acc
+      | s -> CompEntryMap.add_to_list s lr1 acc)
+  CompEntryMap.empty
+  |> inv (CompEntryMap.fold (fun s l acc -> (l, CompEntrySet.elements s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    (pp_brackets @@ pp_list pp_completion_entry)
+    "[])"
+
+let guess_default_value ~mangled_name ~typ =
+  match typ, mangled_name with
+  | _ when EzString.ends_with ~suffix:"option" typ ||
+          EzString.starts_with ~prefix:"option_" mangled_name
+  -> Some "None"
+  | _ when EzString.ends_with ~suffix:"list" typ ||
+  EzString.starts_with ~prefix:"loption_" mangled_name ||
+  EzString.starts_with ~prefix:"list_" mangled_name
+  -> Some "[]"
+  | _ when EzString.ends_with ~suffix:"unit" typ
+  -> Some "()"
+  | _ when EzString.ends_with ~suffix:"bool" typ ||
+  EzString.starts_with ~prefix:"boption_" mangled_name
+  -> Some "false"
+  | _ -> None
+
+let is_nullable_without_recovery = Nonterminal.tabulate begin fun nt ->
+  Nonterminal.nullable nt &&
+    Nonterminal.attributes nt
+    |> List.find_opt (Attribute.has_label "recovery")
+    |> Option.is_none
+  end
+
+let best_guess_default_value = Nonterminal.tabulate begin fun nt ->
+  if not (is_nullable_without_recovery nt)
+  then None
+  else
+  let mangled_name = Nonterminal.mangled_name nt in
+  let guessed_default = Option.bind
+    (Nonterminal.typ nt)
+    (fun typ -> guess_default_value ~mangled_name ~typ) in
+  let given_default = Nonterminal.attributes nt
+    |> List.find_opt (Attribute.has_label "default")
+    |> Option.map Attribute.payload in
+  match given_default with
+    | Some _ -> given_default
+    | None -> guessed_default
+  end
+
+let emit_default_value_of_nullables ppf =
+  Fmt.pf ppf {|
+let guessed_default_value_of_nullables (type a): a Menhir.symbol -> a = function
+  (* If this function does not compile it probably means you're missing a
+     [@default] attribute on a nullable nonterminal in the grammar *)
```

Does this comment still hold (there'e a wildcard with `raise Not_found`) ?

---

I'm also wondering why this function takes a `symbol`, whereas it only makes sense on `nonterminal`.

---

Yes it holds, I put it because on some cases the guessing the type part fails (ex: type is `int * int option` -> defaults to `None` instead of `(0, None)`). I updated the comment for clarity

It returns a symbol for consistency with `Recover.default_value` and to simplify the logic on the `lsp_completion` side of things

---

I believe `symbol` is defines as `type symbol = T of terminal | N of nonterminal`. So passing `(N nt)` to `Recover.default_value` is not an issue. We can postpone that change of API for later if you prefer.